### DISCUSSION TITLE
How to create custom banner?

### DISCUSSION THREAD

**Discussion by noahbald** [nushell/nushell#13197] [Q&A]
I'd like to create a custom banner for my shell, but there's two things I'd like to do which `show_banner` can do but I cannot
- Don't display the banner 
- Display accurate startup time

The [source for show_banner](https://github.com/nushell/nushell/blob/4c82a748c1c9dd4881bca646720e117658ac9c41/crates/nu-std/std/mod.nu#L165) doesn't have anything in it that enabled this behaviour, in fact if I change my config to end with the following...
```nu
$env.config.show_banner = false
use std banner
banner
```
...then running `nu -c "'hello world'"` does show the banner as follows
```
     __  ,
 .--()°'.' Welcome to Nushell,
'|, . ,'   based on the nu language,
 !_-(_\    where all data is structured!

Please join our Discord community at https://discord.gg/NtAbbGn
Our GitHub repository is at https://github.com/nushell/nushell
Our Documentation is located at https://nushell.sh
Tweet us at @nu_shell
Learn how to remove this at: https://nushell.sh/book/configuration.html#remove-welcome-message

It's been this long since Nushell's first commit:
5yrs 1month 10days 13hrs 35mins 56secs 42ms 330µs

Startup Time: -1ns
```
Note the `-1` startup time.

My guess is there's something special [happening here](https://github.com/nushell/nushell/blob/7d2d573eb8bbc14faf37dc3f3bbba658e18e475f/crates/nu-cli/src/repl.rs#L150) that makes this possible. But how can I get this same behaviour as part of my config?

**Comment by fdncred** [comment]
startup time is in `$nu.startup-time` as you can see in the banner custom command https://github.com/nushell/nushell/blob/7d2d573eb8bbc14faf37dc3f3bbba658e18e475f/crates/nu-std/std/mod.nu#L181

I'm not sure why you have a -1 there because you can start nushell with `nu -n --no-std-lib` and still have it. Maybe it's not used when running commands with `-c`.

**Reply by noahbald** [reply]
My guess is startup time isn’t set until config.nu and env.nu are finished running, so it’s not set to a useful value yet.

I thought maybe there’d be a hook we could use for when startup is done, but I couldn’t find anything in the docs.

**Reply by fdncred** [reply]
There is no config.nu nor an env.nu when you run `nu -n --no-std-lib` so I don't think it's that. I think it may be more about the `-c`.

**Reply by noahbald** [reply]
I've figured I can use ` $nu.is-interactive` to hide my banner if I need to.
```nu
# config.nu
if $nu.is-interactive {
  $"
My custom banner
Startup-time: ($nu.startup-time)
"
}
```

> There is no config.nu nor an env.nu when you run nu -n --no-std-lib so I don't think it's that. I think it may be more about the -c.

This isn't really the issue, I only want my custom banner to appear when interactive.

As for showing a startup time from a banner defined in my config, my guess is this won't be possible unless a [hook](https://www.nushell.sh/book/hooks.html) is added for startup completion or something similar.

**Reply by fdncred** [reply]
Now that I think about it, we don't want to move the calculation for $nu.startup-time around because part of startup time is parsing the $nu.config-path and friends. So, if I put that calculation before the config, it wouldn't be an accurate performance metric. I haven't played around with putting it in a hook although I do have an alternate banner that I sometimes call.

```nushell
# show a slightly different banner
def show_banner [] {
    let ellie = [
        "     __  ,"
        " .--()°'.'"
        "'|, . ,'  "
        ' !_-(_\   '
    ]
    let s_mem = (sys mem)
    let s_ho = (sys host)
    print $"(ansi reset)(ansi green)($ellie.0)"
    print $"(ansi green)($ellie.1)  (ansi yellow) (ansi yellow_bold)Nushell (ansi reset)(ansi yellow)v(version | get version)(ansi reset)"
    print $"(ansi green)($ellie.2)  (ansi light_blue) (ansi light_blue_bold)RAM (ansi reset)(ansi light_blue)($s_mem.used) / ($s_mem.total)(ansi reset)"
    print $"(ansi green)($ellie.3)  (ansi light_purple)ﮫ (ansi light_purple_bold)Uptime (ansi reset)(ansi light_purple)($s_ho.uptime)(ansi reset)"
}
```

**Comment by brendon-felix** [comment]
I figured out you can use a `env_change` hook to do this with `HOMEPATH` as the hook variable since it gets set after `$nu.startup-time` is set. Not a perfect solution, but it works for my purposes.

```nushell
# config.nu
$env.config.hooks = {
    env_change: {
        HOMEPATH: [{|| print $"!! Custom banner !!\nstartup time: ($nu.startup-time)"}]
    }
}
```

**Reply by ItsHarper** [reply]
Should be `HOME`, not `HOMEPATH`. Otherwise, this works great, thanks!

Edit: Is there an OS that uses the `HOMEPATH` variable?

**Reply by pWoLiAn** [reply]
I use Windows 11 and HOMEPATH works for me(HOME doesn't)
